### PR TITLE
[EmitC] Remove the forked emitter and generate all the code in the conversion pass

### PR DESCRIPTION
**Pull request by simon-camp**
This resolves the longstanding painpoint of having a custom fork of the EmitC dialect to C emitter infrastructure in IREE. This is done by generating `emitc.verbatim` ops for the missing pieces in the conversion.

Things that need to be addressed:
- Move `createModuleStructure` into a separate file. Skipped for now as it introduces  hundreds lines of diff noise
- The creation of the descriptor arrays are a 1:1 copy from the old CModuleTarget. Helper structs/functions around these would help readability. 
- Convert the `vm.module` to a `builtin.module` in the conversion. (Would need to make the conversion pass run on the outer `builtin.module` and nest the pass manager correctly, I think?)
- The `FuncAnalysis` became kind of a sink where different pieces of information are mashed together with subtle bugs around uninitialized variables from the different constructos.

### CODE REVIEW

**Comment by marbre** [comment]
This is awesome! I take a closer look tomorrow :)

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 1]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
   return success();
 }
 
+/// Generate boilerplate code like includes for the iree C API, include guards,
+/// structures to hold module state, functions and global variables to create a
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: NIT:
```suggestion
/// Generate boilerplate code like includes for the IREE C API, include guards,
/// structures to hold the module state, functions and global variables to create a
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 2]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+                                          "//  __cplusplus");
+    emitc_builders::preprocessorDirective(builder, loc, emitc_builders::ENDIF,
+                                          std::string("//  ") + includeGuard);
+    emitc_builders::preprocessorDirective(builder, loc, emitc_builders::IF,
+                                          "defined(EMITC_IMPLEMENTATION)");
+    builder.create<emitc::IncludeOp>(loc, "iree/vm/ops.h");
+    builder.create<emitc::IncludeOp>(loc, "iree/vm/ops_emitc.h");
+    builder.create<emitc::IncludeOp>(loc, "iree/vm/shims_emitc.h");
+
+    // rodata ops
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: NIT:
```suggestion
    // Rodata ops.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 3]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+                (dependency.isOptional
+                     ? "IREE_VM_MODULE_DEPENDENCY_FLAG_OPTIONAL"
+                     : "IREE_VM_MODULE_DEPENDENCY_FLAG_REQUIRED") +
+                "},";
+      }
+    }
+    deps += "};";
+    builder.create<emitc::VerbatimOp>(loc, deps);
+
+    // imports
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: NIT:
```suggestion
    // Imports.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 4]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+        moduleOp.getOps<IREE::VM::ImportOp>());
+    std::string importName = moduleOp.getName().str() + "_imports_";
+    std::string imports;
+    imports += "static const iree_vm_native_import_descriptor_t " + importName +
+               "[] = {";
+    if (importOps.empty()) {
+      // Empty list placeholder.
+      imports += "{0},";
+    } else {
+      // sort import ops by ordinal
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
      // Sort import ops by ordinal.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 5]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+      }
+    }
+    imports += "};";
+    builder.create<emitc::VerbatimOp>(loc, imports);
+
+    for (auto op : moduleOp.getOps<IREE::VM::ImportOp>()) {
+      opsToRemove.push_back(op);
+    }
+
+    // exports
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
    // Exports.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 6]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+    };
+    std::string exportName = moduleOp.getName().str() + "_exports_";
+    std::string exports;
+    exports += "static const iree_vm_native_export_descriptor_t " + exportName +
+               "[] = {";
+    if (exportedFunctions.empty()) {
+      // Empty list placeholder.
+      exports += "{{0}},";
+    } else {
+      // sort export ops
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
      // Sort export ops.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 7]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+
+        // TODO(simon-camp): support function-level reflection attributes
+        exports += "{" + printStringView(exportName) + ", " +
+                   printStringView(callingConvention) + ", 0, NULL},";
+      }
+    }
+    exports += "};";
+    builder.create<emitc::VerbatimOp>(loc, exports);
+
+    // functions
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
    // Functions.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 8]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+        functions += std::string("{") +
+                     "(iree_vm_native_function_shim_t)iree_emitc_shim, " +
+                     "(iree_vm_native_function_target_t)" + funcName.str() +
+                     "},";
+      }
+    }
+    functions += "};";
+    builder.create<emitc::VerbatimOp>(loc, functions);
+
+    // module descriptor
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
    // Module descriptor.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 9]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+        // exports:
+        + std::to_string(exportedFunctions.size()) + "," + exportName +
+        ","
+        // functions:
+        + std::to_string(exportedFunctions.size()) + "," + functionName + "," +
+        "};";
+
+    builder.create<emitc::VerbatimOp>(loc, descriptor);
+
+    // move functions marked as emitAtEnd to the end of the module
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
    // Move functions marked as `emitAtEnd` to the end of the module.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 10]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+      if (typeConverter.analysis.lookupFunction(func).shouldEmitAtEnd()) {
+        func->moveBefore(moduleOp.getBlock().getTerminator());
+      }
+    }
+
+    builder.setInsertionPoint(moduleOp.getBlock().getTerminator());
+    emitc_builders::preprocessorDirective(builder, loc, emitc_builders::ENDIF,
+                                          "  // EMITC_IMPLEMENTATION");
+
+    // insert a verbatim op before private functions
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
    // Insert a verbatim op before private functions.
```

Comment (marbre) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: This comment states the *what* but maybe you can also add the *why*.

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 11]
```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
...
+
+  for (auto op : opsToRemove) {
+    op->erase();
+  }
+
+  // TODO(simon-camp): The emitc code emitter expects a builtin.module as the
+  // outer container of the supported operations. Instead of nesting a
+  // builtin.module inside the vm.module here we should change this conversion
+  // pass to run on the outer builtin.module and replace the vm.module
+  // completely.
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]:
```suggestion
   // TODO(simon-camp): The Cpp Emitter expects a builtin.module as the
   // outer container of the supported operations. Instead of nesting a
   // builtin.module inside the vm.module as in the current implementation,
   // the conversion pass should be changed to replace the vm.module
   // with a builtin.module.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCBuilders.cpp [hunk 12]
```diff
@@ -7,11 +7,42 @@
 #include "iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCBuilders.h"
 
 #include "llvm/ADT/ArrayRef.h"
-#include "mlir/Dialect/EmitC/IR/EmitC.h"
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/EmitCBuilders.cpp:10]: Even though `mlir/Dialect/EmitC/IR/EmitC.h` is included in `EmitCBuilders.h`, I would keep this include as you create verbatim ops later.

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/VMAnalysis.h [hunk 13]
```diff
@@ -7,97 +7,167 @@
...
 #include "iree/compiler/Dialect/VM/Utils/TypeTable.h"
 #include "mlir/Dialect/EmitC/IR/EmitC.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
 
 namespace mlir::iree_compiler::IREE::VM {
 
+/// TODO(simon-camp): This struct grew from being a wrapper around the
+/// RegisterAllocation and ValueLiveness analyses to also cache other things
+/// needed throughout the conversion. This led to hard to follow failuresin the
+/// usage of this when only part of this struct was correctly initialized. This
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/VMAnalysis.h]:
```suggestion
/// needed throughout the conversion. This led to hard to locate failures
/// when only part of this struct was correctly initialized. This

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir [hunk 14]
```diff
@@ -365,7 +365,7 @@ vm.module @my_module {
 vm.module @my_module {
 
   // CHECK-LABEL: func.func @my_module_call_0v_v_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>)
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir]:
```suggestion
  // TODO: Check for static specifier after converting to `emitc.func`.
  // CHECK-LABEL: func.func @my_module_call_0v_v_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>)
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir [hunk 15]
```diff
@@ -428,7 +428,7 @@ vm.module @my_module {
 
   // CHECK-LABEL: func.func @my_module_call_0iCiD_i_2_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir]:
```suggestion
  // CHECK-LABEL: func.func @my_module_call_0iCiD_i_2_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
  // TODO: Check for static specifier after converting to `emitc.func`.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir [hunk 16]
```diff
@@ -531,7 +531,7 @@ vm.module @my_module {
 
   // CHECK-LABEL: func.func @my_module_call_0iCiD_i_0_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir]:
```suggestion
  // CHECK-LABEL: func.func @my_module_call_0iCiD_i_0_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
  // TODO: Check for static specifier after converting to `emitc.func`.
```

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir [hunk 17]
```diff
@@ -618,7 +618,7 @@ vm.module @my_module {
 
   // CHECK-LABEL: func.func @my_module_call_0r_r_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
```
Comment (marbre) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir]:
```suggestion
  // CHECK-LABEL: func.func @my_module_call_0r_r_import_shim(%arg0: !emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, %arg1: !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>,
  // TODO: Check for static specifier after converting to `emitc.func`.
```

**Review by marbre — CHANGES_REQUESTED** [review]
Mainly checking for what changed in the tests, this look good to me beside some minor suggestions.

However, I was wondering if we should add / register the Func to EmitC conversion patterns with `populateFuncToEmitCPatterns()` as part of this PR. Seems there were multiple `func.func` ops with an `{emitc.static}` attribute (I added TODOs to those). If we mark the `func.func` ops private and apply the conversion, those are automatically emitted as static functions. However, we can also do this in a follow up or even better directly create `emitc.func` ops directly whenever possible. WDYT?

**Comment by simon-camp** [comment]
> Mainly checking for what changed in the tests, this look good to me beside some minor suggestions.
> 
> However, I was wondering if we should add / register the Func to EmitC conversion patterns with `populateFuncToEmitCPatterns()` as part of this PR. Seems there were multiple `func.func` ops with an `{emitc.static}` attribute (I added TODOs to those). If we mark the `func.func` ops private and apply the conversion, those are automatically emitted as static functions. However, we can also do this in a follow up or even better directly create `emitc.func` ops directly whenever possible. WDYT?

Done.

**Comment by marbre** [comment]
> > Mainly checking for what changed in the tests, this look good to me beside some minor suggestions.
> > However, I was wondering if we should add / register the Func to EmitC conversion patterns with `populateFuncToEmitCPatterns()` as part of this PR. Seems there were multiple `func.func` ops with an `{emitc.static}` attribute (I added TODOs to those). If we mark the `func.func` ops private and apply the conversion, those are automatically emitted as static functions. However, we can also do this in a follow up or even better directly create `emitc.func` ops directly whenever possible. WDYT?
> 
> Done.

It was not actually my intention that the switch to `emitc.func` should be part of this PR. I would have kept the PR as small as possible, thus I would have preferred a follow up. Anyway, as the changes are already there you should adjust the PR description.

**Review by benvanik — APPROVED** [review]

**Review by marbre — APPROVED** [review]

**Comment by benvanik** [comment]
you'll need to add a dep on 
```
        "@llvm-project//mlir:ControlFlowDialect",
```

to fix the bazel error:
```
ERROR: /work/compiler/src/iree/compiler/Dialect/VM/Target/C/BUILD.bazel:15:25: Compiling compiler/src/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-17/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer -g0 ... (remaining 176 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
compiler/src/iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp:17:10: error: module //compiler/src/iree/compiler/Dialect/VM/Target/C:C does not depend on a module exporting 'mlir/Dialect/ControlFlow/IR/ControlFlow.h'
   17 | #include "mlir/Dialect/ControlFlow/IR/ControlFlow.h"
      |          ^
```

(and rerun bazel-to-cmake)

**Comment by marbre** [comment]
LGTM, now!
